### PR TITLE
Implements unified implementation of memberOf and subsumes function with 'mapWithPartitionPreview'.

### LINKED ISSUE #190: Remove join from MemberOfFunction
Currently, `MemberOfFunction` works by hashing the input concept column on both the invoking side and the partition side, and then joining the two to form the result.

As we have now demonstrated in `SubsumesFunction`, this is not necessary - we can change the mapper to have a Row type as both input and output, and add the result column while preserving the pre-existing columns.

This will remove the join and potentially improve the performance for datasets with a large number of unique codings.

### PR DESCRIPTION
**Pull request by piotrszul**
Resolves #190.

### CODE REVIEW

**Comment by johngrimes** [comment]
@piotrszul I changed the base of this to `v3`, I assume that's what you meant?

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java [hunk 1]
```diff
@@ -134,19 +124,19 @@ public void memberOfCoding() {
...
+        .withRow("encounter-1", makeEid(0), true)
+        .withRow("encounter-1", makeEid(1), false)
         .withRow("encounter-2", makeEid(0), true)
-        .withRow("encounter-3", makeEid(0), true)
+        .withRow("encounter-3", makeEid(0), false)
         .withRow("encounter-4", makeEid(0), false)
-        .withRow("encounter-5", makeEid(0), false)
+        .withRow("encounter-5", makeEid(0), true)
         .withRow("encounter-6", null, null)
         .build();
```
Comment (johngrimes) [inline fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java:141]: I've been trying to reconcile the differences in these expectations as compared to what is on v3. As far as I can tell the expectations on v3 are wrong, but these are correct...

**Review by johngrimes — CHANGES_REQUESTED** [review]
Thanks @piotrszul, this looks great!

I've made some changes to the build on v3, so I'll merge them back in here now.

Just a few minor issues to fix on this branch, and then I think we are ready to merge!

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java [hunk 2]
```diff
@@ -134,19 +124,19 @@ public void memberOfCoding() {
         Collections.singletonList(argumentExpression));
 
     // Invoke the function.
-    final FhirPath result = new MemberOfFunction(mockCodeMapper).invoke(memberOfInput);
+    final FhirPath result = new MemberOfFunction().invoke(memberOfInput);
 
     // The outcome is somehow random with regard to the sequence passed to MemberOfMapperAnswerer.
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java:129]: This 'randomness' was an issue for the  old version with 'mockCodeMapper' (I believe it was related to unpredicatable ordering of hashes). No longer relevant.

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java (hunk 1 above)
Comment (piotrszul) [reply fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/memberof/MemberOfFunctionTest.java:141]: The expectations in v3 were based on the true/false values returned by the mock so they were probably OK as well. 
The problem with them was though that order of the mock values was different of the order in which they were assigned in the result dataset. 
These now reflect the actual belonging of the ValueSet returned by the mock terminologyServer.

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/subsumes/SubsumptionMapperWithPreviewTest.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    @Override
+    public boolean matches(@Nullable final List<Coding> right) {
+      return right != null &&
+          leftSet.size() == right.size() &&
+          leftSet.equals(right.stream().map(SimpleCoding::new).collect(Collectors.toSet()));
+    }
+
+  }
+
+}
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/subsumes/SubsumptionMapperWithPreviewTest.java:160]: It's tested by SubsumesFunctionTest

File: fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/subsumes/SubsumptionMapperWithPreviewTest.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        terminologyClientFactory, false);
+
+    final List<ImmutablePair<List<SimpleCoding>, List<SimpleCoding>>> inputPairs = Arrays
+        .asList(
+            ImmutablePair.of(Collections.singletonList(CODING1_VERSION1),
+                Collections.singletonList(CODING2_VERSION1)),
+            ImmutablePair.of(Collections.singletonList(CODING1_UNVERSIONED),
+                Collections.singletonList(CODING1_VERSION1)));
+
+    //TODO: check the resulting closure
```
[the remark this answers is not in the document]
Comment (piotrszul) [reply fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/subsumes/SubsumptionMapperWithPreviewTest.java]: Yes, OK. I think should should be tested better as part of #153

**Review by johngrimes — APPROVED** [review]

**Pull request merged**